Whole Body Self Collision Barrier

domrachev03 opened the pull request:
This PR continues a series of PRs, aimed at the introduction of Control Barrier Functions, mentioned in https://github.com/stephane-caron/pink/issues/86. Namely, me and @simeon-ned implemented a whole-body self-collision avoidance barrier, based on Pinocchio's collision model. Namely, it ensures the distance between any collision pair is greater than constant:

$$h(q) = \begin{bmatrix} \ldots \\ d(p^1_i, p^2_i) - d_{min} \\ \ldots \end{bmatrix}
    \quad \forall i \in 0 \ldots N$$

where $N$ is number of collision pairs, $p^k_i$ is the $k-th$ body in $i$-th collision pair,
$d(p^1_i, p^2_i)$ is the distance between collision bodies in the pair,
and $d_{min}$ is minimal distance between any collision bodies.

The functionality is demonstrated on the custom model of two Kuka iiwa14 manipulators:

[Screencast from 06-25-2024 01:20:45 PM.webm](https://github.com/stephane-caron/pink/assets/28687492/8be66364-ea77-4b76-baa5-fede57aec5c7)


There are some notable modifications outside of the barrier class:
1. Barriers now cache the computation of jacobians and barrier values to avoid recomputation.
2. `configuration` now (optionally) accepts collision model and path to an `srdf` file to handle 
collisions. Note that back-compatibility is preserved, and collisions are not computed if the model is not provided
We are kindly asking you to review those changes and make any comments on things you think are done incorrectly.

Note that while it is more expressive than Spherical Barrier, it is also much more computationally expensive. Therefore, we believe that those two barriers have different application scenarios, and should co-exist in this package.

Any further suggestions would be welcome, waiting for your feedback!

P.S. Unit Tests have not been added yet, we want first to make sure we agree on the details of implementation, and then we would quickly add them before merging.

stephane-caron commented on pink/barriers/barrier.py:
```diff
@@ -172,15 +179,23 @@ def compute_qp_objective(
+            self.__q_cache is None
+            or self.__jac_cache is None
+            or self.__barrier_cache is None
+        ) or not np.allclose(self.__q_cache, configuration.q):
```
The function `np.allclose` has relative/absolute tolerance parameters. Depending on the configuration space of the robot, users may want to set them to non-default values.

domrachev03 replied:
I use `np.allclose` with default parameters with the hope that _in practice_ it would return False for any `q_cache != configuration.q`. Also see the discussion below for the logic behind caching.

stephane-caron commented on pink/barriers/barrier.py:
```diff
@@ -208,6 +223,9 @@ def compute_qp_inequality(
         and :math:`\alpha_j` are extended class K functions.
 
+        Note:
+            Jacobian and barrier values are cached to avoid recomputation.
```
I'm not against the idea of caching (a.k.a. memoization), but there are some reasons why it has not been done in Pink so far:

* We assume the robot is moving most of the time.
* The IK should run fast enough for control without memoization (because in the worst case, the robot just moved and we run the full computation anyway).
* It increases code complexity:
    * a bit more code for maintainers to handle 
    * more parameters to future-proof use cases from other users (see `np.allclose` below)
    * more logic for future developers to keep in mind (e.g. someone developing a new barrier may will have to know this to understand why their custom barrier/Jacobian functions are not called all the time)

Have you measured an improvement from caching in your tests so far?

domrachev03 replied:
The main reason why we've decided to implement caching is that we are generally computing both cost and constraint for the same iteration, which results in computing Jacobian and bias twice per iteration. 

This was particularly severe, as the controller struggled to keep up even at 50Hz. So, this was the main purpose of caching: not to skip updates between pink's iterations, but to skip updates during one pink iteration.

domrachev03 replied:
Probably this is not the best solution possible, but I failed to come up with something more descent without changing the logic of `solve_ik`.

Any suggestions on this matter are welcome!

stephane-caron commented on pink/barriers/self_collision_barrier.py:
```diff
@@ -0,0 +1,196 @@
+    def compute_jacobian(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the Jacobian matrix of the barrier function.
+
+        The jacobian matrix could be represented as stacked gradients of each
```
nit (there are a few other occurrences of this below):

```suggestion
        The Jacobian matrix could be represented as stacked gradients of each
```

stephane-caron commented on pink/configuration.py:
```diff
@@ -89,12 +105,28 @@ def __init__(
+        self.collision_model = collision_model
+        self.update_collision = self.collision_model is not None
+
+        if self.collision_model is not None:
```
This looks like work users should do when they prepare their collision models, rather than relying on the Configuration constructor to do it internally.

* We could for instance move it to a utility function in `utils.py` with a prototype like `process_collision_pairs(collision_model, srdf_path)`.
* And/or (independent from Pink) we could process collision pairs automatically in robot_descriptions' [`load_in_pinocchio`](https://github.com/robot-descriptions/robot_descriptions.py/blob/943ad0512ad63043f5f774fe141715cc4873e20b/robot_descriptions/loaders/pinocchio.py#L33) function, doing something like "if the description has an `SRDF_PATH`, then process its collision pairs".

domrachev03 replied:
I anticipate the second approach, since (at least for me) a `.srdf` file is subject to a frequent dynamic change, since a set of excluded pairs strongly depends on the performed task, and might be tweaked to find suitable performance/accuracy trade-off.

With that being said, I find the first approach suitable, and would implement it, hope I got your idea correctly

stephane-caron commented on pink/configuration.py:
```diff
@@ -89,12 +105,28 @@ def __init__(
     def update(self, q: Optional[np.ndarray] = None) -> None:
-        """Update configuration to a new vector and run forward kinematics.
+        """Update configuration to a new vector and run forward kinematics and
+        collision pairs distance calculations, if specified.
```
I think the linter will complain that the brief function description is on two lines? :thinking:

domrachev03 replied:
It seems that all github actions are running for a PR, therefore I suppose this is fine for the linter, dunno ¯\_(ツ)_/¯

stephane-caron commented on pink/configuration.py:
```diff
@@ -103,6 +135,25 @@ def update(self, q: Optional[np.ndarray] = None) -> None:
             self.q = q_readonly
+
+        # Compute collisions, if needed
+        if self.update_collision:
```
We can do an equivalent check without the additional `update_collision` attribute:

```suggestion
        if self.collision_model is not None:
```

Unless we want users to see them (e.g. if it were convenient for users to set ``configuration.update_collision = False`` at runtime, but here I guess that would be a bad idea), it is better to have less public attributes in objects.

stephane-caron reviewed (commented):
Thank you for this excellent proposal :smiley: The gist of this PR, that is the barrier and its Jacobian, seem OK. The questions/improvements I see are more on the software side (URDF distribution, caching and SRDF processing).

> Therefore, we believe that those two barriers have different application scenarios, and should co-exist in this package.

:+1: 

> P.S. Unit Tests have not been added yet, we want first to make sure we agree on the details of implementation, and then we would quickly add them before merging.

Sounds good!

domrachev03 commented on pink/barriers/barrier.py:
```diff
@@ -172,15 +179,23 @@ def compute_qp_objective(
                 configuration
             )
-            jac_squared_norm = np.linalg.norm(jac) ** 2
+            jac_squared_norm = np.sum(self.__jac_cache**2)
```
Indeed, this is a surprise for me, it becomes even worse when dimensions grow. 

Thank you very much for noting this, I would be more careful about such reasoning in the future.

domrachev03 replied:
This is very surprising for me that it works in this way.

Thank you for your attention to details, I would pay more attention to those details.

stephane-caron replied:
No worries. Here I believe we see this because `np.linalg` will run a single NumPy function over the array, while the `np.sum` variant will first create a whole array for the squared Jacobian before calling the NumPy sum function.

domrachev03 commented on pink/barriers/barrier.py:
```diff
@@ -216,8 +234,17 @@ def compute_qp_inequality(
+            self.__barrier_cache = self.compute_barrier(configuration).copy()
+
+        G = -self.__jac_cache.copy() / dt
+        barrier_value = self.__barrier_cache.copy()
```
Tbh, I have a nasty habit of copying everything and everywhere to avoid those rare but very tricky and disturbing errors with views.

I've double-checked this piece of code, and I believe that it's not necessary at all in both `compute_qp_objective` and `compute_qp_inequality`

domrachev03 commented on examples/barriers/kukas_self_collision.py:
```diff
@@ -0,0 +1,131 @@
+if __name__ == "__main__":
+    # Load the robot and define the custom frames
+    urdf_path = "examples/barriers/models/iiwa14_spheres_collision.urdf"
+    srdf_path = "examples/barriers/models/iiwa14_spheres_collision.srdf"
```
I need some help there because I'm not sure what is the best way to represent this.

The thing is, this is a copy of the `iiwa14` model from `robot_descriptions.py` with the following changes:
1. Base joint collision geometry is changed from cylinder to sphere.
2. The manipulators are doubled, and all joints and frames are renamed to `first_X` and `second_X`.
3. The `pdf` file is added (but I consider it more like a user configuration, than a part of the general-purpose model.

With that being said, I'm not sure it should be a separate entry in `robot_menagerie.py`. I've looked for other ways to create this model, the ones I've considered are:
1. Use `pin.appendModel(...)`. But I'm not sure how it handles two models with practically similar links and frame names.
2. Leave only `.urdf` and `.srdf` files in the example folder, and load all the meshes from `robot_description.py`/other repository. However, I don't know how to make it properly.

If, given the fact that this is practically a clone of `robot_menagerie.py` repository, you still think it's fine to add it there, I am 100% ready to do it. Any suggestions on other approaches to handle this are welcome too.

stephane-caron replied:
I agree with @simeon-ned's suggestion to use `pin.appendModel` for the dual-arm part. Before appending you can update the collision model of the arm in Python to replace the cylinders by spheres (NB: make sure to recreate the collision data associated with the collision model after updating it).

domrachev03 commented:
Thank you once again for such detailed feedback. I'm really glad to see it since I learned something new about good practices in coding!

I believe I've addressed the majority of your comments, the only remaining questions are:
1. Is there a better way to avoid recalculation of Jacobian and bias while computing the barrier's objective and constraints?
2. How to include a model with two iiwas in the example? Should we add it to the `robot_descriptions.py`?
3. Add unit testing (already wip).

Thank you once again for your detailed feedback, waiting for your opinion on first two questions.

simeon-ned commented:
> Thank you once again for such detailed feedback. I'm really glad to see it since I learned something new about good practices in coding!
> 
> I believe I've addressed the majority of your comments, the only remaining questions are:
> 
> 1. Is there a better way to avoid recalculation of Jacobian and bias while computing the barrier's objective and constraints?
> 2. How to include a model with two iiwas in the example? Should we add it to the `robot_descriptions.py`?
> 3. Add unit testing (already wip).
> 
> Thank you once again for your detailed feedback, waiting for your opinion on first two questions.
It seems for me that we can make the double arm model through: `pin.appendModel` in similar way it is done in [flying_dual_arm example](https://github.com/stephane-caron/pink/blob/main/examples/flying_dual_arm_ur3.py)

domrachev03 commented:
> It seems for me that we can make the double arm model through: `pin.appendModel` in similar way it is done in [flying_dual_arm example](https://github.com/stephane-caron/pink/blob/main/examples/flying_dual_arm_ur3.py)

Okay, now I see that it is quite simple to change frame names... I'll then rewrite the example in this way and remove model from the example. Then the only question is left is regarding jacobian and bias caching.

domrachev03 commented:
Sorry for the absence, we've rewritten an example as you've suggested. Now only srdf file left, which represents which collision pairs should be removed.

Additionally, we've implemented the tests, which cover whole new functionality.

To sum up, we've implemented all discussed fixes, rewrote an example and added tests. Waiting for your feedback!

domrachev03 commented:
@stephane-caron I think everything is ready from our side to merge, so if you see no problems, I think we could merge those barriers. What do you think?

stephane-caron commented on pink/configuration.py:
```diff
@@ -45,11 +51,16 @@ class Configuration:
     Attributes:
         data: Data corresponding to :data:`Configuration.model`.
         model: Kinodynamic model.
+        update_collision: If true (default), calculate collisions in the models.
```
```suggestion
```

stephane-caron commented on examples/barriers/README.md:
```diff
@@ -60,3 +60,21 @@ https://github.com/domrachev03/pink/assets/28687492/f8c4bc8d-63e3-4bf7-a34f-e7ed
+
+| Barrier | Gain |
+|------|------|
+| Self Collision Avoidance | $10$ 
```
```suggestion
| Self Collision Avoidance | $10$ |
```

stephane-caron commented:
My apologies for taking longer than usual on this PR (due in part to attending a conference last week). The PR is in excellent condition, let's merge.

Thank you for this excellent work :star_struck:
